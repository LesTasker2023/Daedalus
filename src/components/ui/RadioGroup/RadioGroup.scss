// RadioGroup component SCSS overrides
// Use this file for ENHANCEMENTS that can't be achieved with Tailwind classes
// Avoid duplicating what Tailwind already provides

.radiogroup-wrapper {
  // Enhanced focus styles for better accessibility
  .radiogroup--loading {
    opacity: 0.7;
    pointer-events: none;

    &::after {
      content: "";
      position: absolute;
      top: 50%;
      left: 50%;
      width: 16px;
      height: 16px;
      margin: -8px 0 0 -8px;
      border: 2px solid transparent;
      border-top-color: hsl(var(--primary));
      border-radius: 50%;
      animation: radiogroup-spin 1s linear infinite;
    }
  }

  // Error state styling (STANDARDIZED ADDITION)
  &--error {
    .radiogroup-item {
      border-color: hsl(var(--destructive));

      &:focus-visible {
        outline-color: hsl(var(--destructive));
        box-shadow: 0 0 0 2px hsl(var(--destructive) / 0.2);
      }

      &[data-state="checked"] {
        border-color: hsl(var(--destructive));
        background-color: hsl(var(--destructive));
      }
    }
  }
}

@keyframes radiogroup-spin {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}

// Enhanced radio item styles
.radiogroup-item {
  // Enhanced focus styles for better accessibility
  &:focus-visible {
    outline: 2px solid hsl(var(--ring));
    outline-offset: 2px;
  }

  // Smooth transitions
  transition: all 0.2s ease-in-out;

  // Enhanced hover states
  &:hover:not(:disabled) {
    border-color: hsl(var(--primary));
    box-shadow: 0 0 0 1px hsl(var(--primary) / 0.3);
  }

  // Loading state
  &--loading {
    opacity: 0.5;
    cursor: not-allowed;
  }
}

// High contrast mode support
@media (prefers-contrast: high) {
  .radiogroup-item {
    border-width: 2px;

    &:focus-visible {
      outline-width: 3px;
    }
  }
}

// Reduced motion support
@media (prefers-reduced-motion: reduce) {
  .radiogroup--loading::after {
    animation: none;
  }

  .radiogroup-item {
    transition: none;
  }
}
