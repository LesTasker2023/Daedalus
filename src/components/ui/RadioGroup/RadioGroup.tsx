'use client';

import * as RadioGroupPrimitive from '@radix-ui/react-radio-group';
import { Circle } from 'lucide-react';
import * as React from 'react';

import { cn } from '@/lib/utils';

export interface RadioGroupProps
  extends React.ComponentPropsWithoutRef<typeof RadioGroupPrimitive.Root> {
  label?: string;
  error?: string;
  helperText?: string;
  loading?: boolean;
  required?: boolean;
}

const RadioGroup = React.forwardRef<
  React.ElementRef<typeof RadioGroupPrimitive.Root>,
  RadioGroupProps
>(
  (
    { className, label, error, helperText, loading, required, disabled, children, ...props },
    ref
  ) => {
    const groupId = React.useId();

    return (
      <div className="space-y-2">
        {label && (
          <label
            htmlFor={groupId}
            className={cn(
              'text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70',
              required && "after:content-['*'] after:ml-0.5 after:text-destructive",
              error && 'text-destructive'
            )}
          >
            {label}
          </label>
        )}
        <RadioGroupPrimitive.Root
          id={groupId}
          className={cn('grid gap-2', className)}
          disabled={disabled || loading}
          ref={ref}
          {...props}
        >
          {children}
        </RadioGroupPrimitive.Root>
        {helperText && !error && <p className="text-sm text-muted-foreground">{helperText}</p>}
        {error && <p className="text-sm text-destructive">{error}</p>}
        {loading && <p className="text-sm text-muted-foreground">Loading...</p>}
      </div>
    );
  }
);

RadioGroup.displayName = RadioGroupPrimitive.Root.displayName;

const RadioGroupItem = React.forwardRef<
  React.ElementRef<typeof RadioGroupPrimitive.Item>,
  React.ComponentPropsWithoutRef<typeof RadioGroupPrimitive.Item>
>(({ className, ...props }, ref) => {
  return (
    <RadioGroupPrimitive.Item
      ref={ref}
      className={cn(
        'aspect-square h-4 w-4 rounded-full border border-primary text-primary ring-offset-background focus:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50',
        className
      )}
      {...props}
    >
      <RadioGroupPrimitive.Indicator className="flex items-center justify-center">
        <Circle className="h-2.5 w-2.5 fill-current text-current" />
      </RadioGroupPrimitive.Indicator>
    </RadioGroupPrimitive.Item>
  );
});

RadioGroupItem.displayName = RadioGroupPrimitive.Item.displayName;

export { RadioGroup, RadioGroupItem };
