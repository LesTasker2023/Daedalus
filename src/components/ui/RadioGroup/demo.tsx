'use client';

import React, { useState } from 'react';
import { RadioGroup } from '.';

export default function RadioGroupDemo() {
  const [state, setState] = useState(false);

  return (
    <div className="space-y-6 p-6 max-w-2xl mx-auto">
      <div className="space-y-2">
        <h1 className="text-2xl font-bold">RadioGroup Component</h1>
        <p className="text-muted-foreground">Single-choice input groups</p>
      </div>

      <div className="space-y-4">
        <div className="space-y-2">
          <h3 className="text-lg font-semibold">Basic Usage</h3>
          <div className="border rounded-lg p-4 bg-background">
            <RadioGroup>RadioGroup Example</RadioGroup>
          </div>
        </div>

        <div className="space-y-2">
          <h3 className="text-lg font-semibold">Interactive</h3>
          <div className="border rounded-lg p-4 bg-background space-y-2">
            <RadioGroup onClick={() => setState(!state)}>
              Click me {state ? '(Active)' : '(Inactive)'}
            </RadioGroup>
            <p className="text-sm text-muted-foreground">State: {state ? 'Active' : 'Inactive'}</p>
          </div>
        </div>
      </div>
    </div>
  );
}
