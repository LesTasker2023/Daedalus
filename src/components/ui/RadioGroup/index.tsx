export { default } from './radio-group';
export { RadioGroupItem } from './radio-group';
export type { RadioGroupProps } from './radio-group';
