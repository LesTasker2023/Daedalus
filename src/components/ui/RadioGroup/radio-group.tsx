'use client';

import * as RadioGroupPrimitive from '@radix-ui/react-radio-group';
import { Circle } from 'lucide-react';
import * as React from 'react';

import { cn } from '@/lib/utils';
import './RadioGroup.scss';

// Enhanced RadioGroup with additional props
export interface RadioGroupProps
  extends React.ComponentPropsWithoutRef<typeof RadioGroupPrimitive.Root> {
  label?: string;
  error?: string;
  helperText?: string;
  loading?: boolean;
  required?: boolean;
}

const RadioGroup = React.forwardRef<
  React.ElementRef<typeof RadioGroupPrimitive.Root>,
  RadioGroupProps
>(
  (
    { className, label, error, helperText, loading = false, required = false, disabled, ...props },
    ref
  ) => {
    // Smart conditional rendering - only wrap when enhanced features are needed
    const radioGroupElement = (
      <RadioGroupPrimitive.Root
        className={cn(
          'grid gap-2',
          loading && 'radiogroup--loading',
          error && 'radiogroup--error',
          className
        )}
        disabled={disabled || loading}
        {...props}
        ref={ref}
      />
    );

    // Return minimal DOM when no enhanced features
    if (!label && !error && !helperText) {
      return radioGroupElement;
    }

    // Enhanced wrapper with label, error, and helper text
    return (
      <div className="radiogroup-wrapper space-y-2">
        {label && (
          <label
            className={cn(
              'text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70',
              required && "after:content-['*'] after:ml-0.5 after:text-destructive",
              error && 'text-destructive'
            )}
          >
            {label}
          </label>
        )}
        {radioGroupElement}
        {(error || helperText) && (
          <p className={cn('text-sm', error ? 'text-destructive' : 'text-muted-foreground')}>
            {error || helperText}
          </p>
        )}
      </div>
    );
  }
);
RadioGroup.displayName = RadioGroupPrimitive.Root.displayName;

const RadioGroupItem = React.forwardRef<
  React.ElementRef<typeof RadioGroupPrimitive.Item>,
  React.ComponentPropsWithoutRef<typeof RadioGroupPrimitive.Item>
>(({ className, ...props }, ref) => {
  return (
    <RadioGroupPrimitive.Item
      ref={ref}
      className={cn(
        'aspect-square h-4 w-4 rounded-full border border-primary text-primary shadow focus:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50',
        className
      )}
      {...props}
    >
      <RadioGroupPrimitive.Indicator className="flex items-center justify-center">
        <Circle className="h-3.5 w-3.5 fill-primary" />
      </RadioGroupPrimitive.Indicator>
    </RadioGroupPrimitive.Item>
  );
});
RadioGroupItem.displayName = RadioGroupPrimitive.Item.displayName;

export default RadioGroup;
export { RadioGroupItem };
