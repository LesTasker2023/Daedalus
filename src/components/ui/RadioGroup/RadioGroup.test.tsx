import { render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import RadioGroup, { RadioGroupItem } from './radio-group';

describe('RadioGroup', () => {
  describe('Snapshots', () => {
    it('matches default snapshot', () => {
      const { container } = render(
        <RadioGroup defaultValue="option1">
          <div className="flex items-center space-x-2">
            <RadioGroupItem value="option1" id="option1" />
            <label htmlFor="option1">Option 1</label>
          </div>
          <div className="flex items-center space-x-2">
            <RadioGroupItem value="option2" id="option2" />
            <label htmlFor="option2">Option 2</label>
          </div>
        </RadioGroup>
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Functionality', () => {
    it('renders correctly', () => {
      render(
        <RadioGroup data-testid="radio-group">
          <RadioGroupItem value="test" />
        </RadioGroup>
      );
      expect(screen.getByTestId('radio-group')).toBeInTheDocument();
    });
  });
});
