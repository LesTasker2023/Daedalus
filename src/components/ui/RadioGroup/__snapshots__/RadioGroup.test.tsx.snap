// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RadioGroup > Snapshots > matches default snapshot 1`] = `
<div
  aria-required="false"
  class="grid gap-2"
  dir="ltr"
  role="radiogroup"
  style="outline: none;"
  tabindex="0"
>
  <div
    class="flex items-center space-x-2"
  >
    <button
      aria-checked="true"
      class="aspect-square h-4 w-4 rounded-full border border-primary text-primary shadow focus:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50"
      data-radix-collection-item=""
      data-state="checked"
      id="option1"
      role="radio"
      tabindex="-1"
      type="button"
      value="option1"
    >
      <span
        class="flex items-center justify-center"
        data-state="checked"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-circle h-3.5 w-3.5 fill-primary"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="12"
            cy="12"
            r="10"
          />
        </svg>
      </span>
    </button>
    <label
      for="option1"
    >
      Option 1
    </label>
  </div>
  <div
    class="flex items-center space-x-2"
  >
    <button
      aria-checked="false"
      class="aspect-square h-4 w-4 rounded-full border border-primary text-primary shadow focus:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50"
      data-radix-collection-item=""
      data-state="unchecked"
      id="option2"
      role="radio"
      tabindex="-1"
      type="button"
      value="option2"
    />
    <label
      for="option2"
    >
      Option 2
    </label>
  </div>
</div>
`;
